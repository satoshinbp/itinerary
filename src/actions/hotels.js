export const addHotel = (
  {
    id='',
    tripId = '',
    name = '',
    checkInDate = 0,
    checkOutDate = 0,
    checkInTime = 0,
    checkOutTime = 0,
    ETA = 0,
    ETD = 0,
    location ='',
    note = ''
  } = {}
) => ({
  type: 'ADD_HOTEL',
  hotel: {
    id,
    tripId,
    name,
    checkInDate,
    checkOutDate,
    checkInTime,
    checkOutTime,
    ETA,
    ETD,
    location,
    note
  }
})

export const editHotel = (id, updates) => ({
  type: 'EDIT_HOTEL',
  id,
  updates
})

export const removeHotel = (id = '') => ({
  type: 'REMOVE_HOTEL',
  id
})