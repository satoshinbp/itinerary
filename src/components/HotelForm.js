import React, { useState } from 'react'
import { DateRangePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'
import DateFnsUtils from '@date-io/date-fns'
import { TimePicker, MuiPickersUtilsProvider } from '@material-ui/pickers'
import moment from 'moment'

export default (props) => {
  const id = props.hotel ? props.hotel.id : props.id
  const tripId = props.hotel ? props.hotel.tripId : props.tripId
  const [name, setName] = useState(props.hotel ? props.hotel.name : '')
  const [checkInDate, setCheckInDate] = useState(props.hotel ? moment(props.hotel.checkInDate) : moment(props.date))
  const [checkOutDate, setCheckOutDate] = useState(props.hotel ? moment(props.hotel.checkOutDate) : moment(props.date).add(1, 'day'))
  const [checkInTime, setCheckInTime] = useState(props.hotel ? moment(props.hotel.checkInTime) : moment("1500", "hmm"))
  const [checkOutTime, setCheckOutTime] = useState(props.hotel ? moment(props.hotel.checkOutTime) : moment("1100", "hmm"))
  const [ETA, setETA] = useState(props.hotel ? props.hotel.ETA : moment("1500", "hmm"))
  const [ETD, setETD] = useState(props.hotel ? props.hotel.ETD : moment("1100", "hmm"))
  const [location, setLocation] = useState(props.hotel ? props.hotel.location : '')
  const [note, setNote] = useState(props.hotel ? props.hotel.note : '')
  const [focusedInput, setFocusedInput] = useState(null)
  const [error, setError] = useState('')

  let numberOfMonths
  if (window.matchMedia("(min-width: 45rem)").matches) {
    numberOfMonths = 2
  } else {
    numberOfMonths = 1
  }
  
  const onDatesChange = ({ startDate, endDate }) => {
    if (startDate) {
      setCheckInDate(moment(startDate).startOf('day'))
    }
    if (endDate) {
      setCheckOutDate(moment(endDate).startOf('day'))
    }
  }
  const onSubmit = e => {
    e.preventDefault()

    if (!name) {
      setError('Please provide hotel name.')
    } else {
      setError('')
      props.onSubmit({ id,
        tripId,
        name,
        checkInDate: checkInDate.valueOf(),
        checkOutDate: checkOutDate.valueOf(),
        checkInTime: checkInTime.valueOf(),
        checkOutTime: checkOutTime.valueOf(),
        ETA: ETA.valueOf(),
        ETD: ETD.valueOf(),
        location,
        note
      })
    }
  }
  
  return (
    <form className="form" onSubmit={onSubmit}>
      {error && <p className="form__error">{error}</p>}
      <input
        type="text"
        placeholder="name"
        autoFocus
        className="text-input"
        value={name}
        onChange={e => setName(e.target.value)}
      />
      <div className="date-time-picker">
        <DateRangePicker
          numberOfMonths={numberOfMonths}
          startDate={checkInDate}
          startDateId="check_in_date_id"
          endDate={checkOutDate}
          endDateId="check_out_date_id"
          onDatesChange={onDatesChange}
          focusedInput={focusedInput}
          onFocusChange={focusedInput => setFocusedInput(focusedInput)}
          isOutsideRange={() => false}
          minimumNights={1}
          displayFormat="YYYY/MM/DD"
        />
      </div>
      <div className="date-time-picker">
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={checkInTime} onChange={time => setCheckInTime(moment(time))} label="Check-in Time" />
        </MuiPickersUtilsProvider>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={checkOutTime} onChange={time => setCheckOutTime(moment(time))} label="Check-out Time" />
        </MuiPickersUtilsProvider>
      </div>
      <div className="date-time-picker">
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={ETA} onChange={time => setETA(moment(time))} label="ETA" />
        </MuiPickersUtilsProvider>
        <MuiPickersUtilsProvider utils={DateFnsUtils}>
          <TimePicker value={ETD} onChange={time => setETD(moment(time))} label="ETD" />
        </MuiPickersUtilsProvider>
      </div>
      <input
        type="text"
        placeholder="location"
        className="text-input"
        value={location}
        onChange={e => setLocation(e.target.value)}
      />
      <textarea
        placeholder="Add a note for your hotel (optional)"
        className="textarea"
        value={note}
        onChange={e => setNote(e.target.value)}
      >
      </textarea>
      <button className="btn-push">Save Hotel</button>
    </form>
  )
}