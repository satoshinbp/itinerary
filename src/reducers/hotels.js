const hotelsReducerDefaultState = []

export default (state = hotelsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_HOTEL':
      return [
        ...state,
        action.hotel
      ]
    case 'EDIT_HOTEL':
      return state.map((hotel) => {
        if (hotel.id === action.id) {
          return {
            ...hotel,
            ...action.updates
          }
        } else {
          return hotel
        }
      })
    case 'REMOVE_HOTEL':
      return state.filter(({ id }) => id !== action.id)
    default:
      return state
  }
}