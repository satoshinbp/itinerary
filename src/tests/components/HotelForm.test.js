import React from 'react'
import { shallow } from 'enzyme'
import moment from 'moment'
import HotelForm from '../../components/HotelForm'
import hotels from '../fixtures/hotels'

Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation(query => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  }))
})

let setState, useStateSpy

beforeEach(() => {
  setState = jest.fn()
  useStateSpy = jest.spyOn(React, 'useState')
  useStateSpy.mockImplementation(init => [init, setState])
})

test('should render HotelForm correctly', () => {
  const wrapper = shallow(<HotelForm />)
  expect(wrapper).toMatchSnapshot()
})

test('should render HotelForm correctly with trip data', () => {
  const wrapper = shallow(<HotelForm hotel={hotels[0]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<HotelForm />)
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  })
  expect(setState).toBeCalledWith(expect.any(String))
})

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn()
  const wrapper = shallow(<HotelForm onSubmit={onSubmitSpy} hotel={hotels[0]} />)
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  })
  expect(setState).toBeCalledWith('')
  expect(onSubmitSpy).lastCalledWith({
    id: hotels[0].id,
    tripId: hotels[0].tripId,
    name: hotels[0].name,
    checkInDate: hotels[0].checkInDate,
    checkOutDate: hotels[0].checkOutDate,
    checkInTime: hotels[0].checkInTime,
    checkOutTime: hotels[0].checkOutTime,
    ETA: hotels[0].ETA,
    ETD: hotels[0].ETD,
    location: hotels[0].location,
    note: hotels[0].note
  })
})

test('should set name on input change', () => {
  const value = 'New Name'
  const wrapper = shallow(<HotelForm />)
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set location on input change', () => {
  const value = 'New Location'
  const wrapper = shallow(<HotelForm />)
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set note on textarea change', () => {
  const value = 'New Note'
  const wrapper = shallow(<HotelForm />)
  wrapper.find('textarea').simulate('change', {
    target: { value }
  })
  expect(setState).toBeCalledWith(value)
})

test('should set date on date change', () => {
  const startDate = moment(0)
  const endDate = moment(1000)
  const wrapper = shallow(<HotelForm />)
  wrapper.find('withStyles(DateRangePicker)').prop('onDatesChange')({ startDate, endDate })
  expect(setState).toBeCalledWith(moment(startDate).startOf('day'))
  expect(setState).toBeCalledWith(moment(endDate).startOf('day'))
})

test('should set calendar focus on change', () => {
  const focusedInput = 'startDate'
  const wrapper = shallow(<HotelForm />)
  wrapper.find('withStyles(DateRangePicker)').prop('onFocusChange')(focusedInput)
  expect(setState).toBeCalledWith(focusedInput)
})

test('should set check-in time on change', () => {
  const time = moment().hour(18).minutes(0)
  const wrapper = shallow(<HotelForm />)
  wrapper.find('PickerWithState').at(0).prop('onChange')(time)
  expect(setState).toBeCalledWith(time)
})

test('should set check-out time on change', () => {
  const time = moment().hour(9).minutes(0)
  const wrapper = shallow(<HotelForm />)
  wrapper.find('PickerWithState').at(1).prop('onChange')(time)
  expect(setState).toBeCalledWith(time)
})

test('should set ETA on change', () => {
  const time = moment().hour(17).minutes(0)
  const wrapper = shallow(<HotelForm />)
  wrapper.find('PickerWithState').at(2).prop('onChange')(time)
  expect(setState).toBeCalledWith(time)
})

test('should set ETD on change', () => {
  const time = moment().hour(10).minutes(0)
  const wrapper = shallow(<HotelForm />)
  wrapper.find('PickerWithState').at(3).prop('onChange')(time)
  expect(setState).toBeCalledWith(time)
})