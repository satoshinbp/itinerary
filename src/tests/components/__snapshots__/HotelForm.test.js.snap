// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render HotelForm correctly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="name"
    type="text"
    value=""
  />
  <div
    className="date-time-picker"
  >
    <withStyles(DateRangePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customArrowIcon={null}
      customCloseIcon={null}
      customInputIcon={null}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="YYYY/MM/DD"
      enableOutsideDays={false}
      endDate={"1970-01-02T00:00:00.000Z"}
      endDateId="check_out_date_id"
      endDatePlaceholderText="End Date"
      firstDayOfWeek={null}
      focusedInput={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      minimumNights={1}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDatesChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableEndDate": [Function],
          "chooseAvailableStartDate": [Function],
          "clearDates": "Clear Dates",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsSelectedAsEndDate": [Function],
          "dateIsSelectedAsStartDate": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDates={false}
      required={false}
      screenReaderInputMessage=""
      showClearDates={false}
      showDefaultInputIcon={false}
      small={false}
      startDate={"1970-01-01T00:00:00.000Z"}
      startDateId="check_in_date_id"
      startDatePlaceholderText="Start Date"
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Check-in Time"
        onChange={[Function]}
        openTo="hours"
        value={"1500-01-01T00:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Check-out Time"
        onChange={[Function]}
        openTo="hours"
        value={"1100-01-01T00:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="ETA"
        onChange={[Function]}
        openTo="hours"
        value={"1500-01-01T00:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="ETD"
        onChange={[Function]}
        openTo="hours"
        value={"1100-01-01T00:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="location"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your hotel (optional)"
    value=""
  />
  <button
    className="btn-push"
  >
    Save Hotel
  </button>
</form>
`;

exports[`should render HotelForm correctly with trip data 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="name"
    type="text"
    value="Waikiki Resort Hotel"
  />
  <div
    className="date-time-picker"
  >
    <withStyles(DateRangePicker)
      anchorDirection="left"
      appendToBody={false}
      block={false}
      calendarInfoPosition="bottom"
      customArrowIcon={null}
      customCloseIcon={null}
      customInputIcon={null}
      dayPickerNavigationInlineStyles={null}
      daySize={39}
      disableScroll={false}
      disabled={false}
      displayFormat="YYYY/MM/DD"
      enableOutsideDays={false}
      endDate={"1970-01-03T00:00:00.000Z"}
      endDateId="check_out_date_id"
      endDatePlaceholderText="End Date"
      firstDayOfWeek={null}
      focusedInput={null}
      hideKeyboardShortcutsPanel={false}
      horizontalMargin={0}
      initialVisibleMonth={null}
      inputIconPosition="before"
      isDayBlocked={[Function]}
      isDayHighlighted={[Function]}
      isOutsideRange={[Function]}
      isRTL={false}
      keepFocusOnInput={false}
      keepOpenOnDateSelect={false}
      minimumNights={1}
      monthFormat="MMMM YYYY"
      navNext={null}
      navPosition="navPositionTop"
      navPrev={null}
      noBorder={false}
      numberOfMonths={1}
      onClose={[Function]}
      onDatesChange={[Function]}
      onFocusChange={[Function]}
      onNextMonthClick={[Function]}
      onPrevMonthClick={[Function]}
      openDirection="down"
      orientation="horizontal"
      phrases={
        Object {
          "calendarLabel": "Calendar",
          "chooseAvailableEndDate": [Function],
          "chooseAvailableStartDate": [Function],
          "clearDates": "Clear Dates",
          "closeDatePicker": "Close",
          "dateIsSelected": [Function],
          "dateIsSelectedAsEndDate": [Function],
          "dateIsSelectedAsStartDate": [Function],
          "dateIsUnavailable": [Function],
          "enterKey": "Enter key",
          "escape": "Escape key",
          "focusStartDate": "Interact with the calendar and add the check-in date for your trip.",
          "hideKeyboardShortcutsPanel": "Close the shortcuts panel.",
          "homeEnd": "Home and end keys",
          "jumpToNextMonth": "Move forward to switch to the next month.",
          "jumpToPrevMonth": "Move backward to switch to the previous month.",
          "keyboardBackwardNavigationInstructions": "Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardForwardNavigationInstructions": "Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
          "keyboardShortcuts": "Keyboard Shortcuts",
          "leftArrowRightArrow": "Right and left arrow keys",
          "moveFocusByOneDay": "Move backward (left) and forward (right) by one day.",
          "moveFocusByOneMonth": "Switch months.",
          "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week.",
          "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week.",
          "openThisPanel": "Open this panel.",
          "pageUpPageDown": "page up and page down keys",
          "questionMark": "Question mark",
          "returnFocusToInput": "Return to the date input field.",
          "roleDescription": "datepicker",
          "selectFocusedDate": "Select the date in focus.",
          "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel.",
          "upArrowDownArrow": "up and down arrow keys",
        }
      }
      readOnly={false}
      regular={false}
      renderCalendarInfo={null}
      renderDayContents={null}
      renderMonthElement={null}
      renderMonthText={null}
      renderNavNextButton={null}
      renderNavPrevButton={null}
      renderWeekHeaderElement={null}
      reopenPickerOnClearDates={false}
      required={false}
      screenReaderInputMessage=""
      showClearDates={false}
      showDefaultInputIcon={false}
      small={false}
      startDate={"1970-01-02T00:00:00.000Z"}
      startDateId="check_in_date_id"
      startDatePlaceholderText="Start Date"
      verticalHeight={null}
      verticalSpacing={22}
      weekDayFormat="dd"
      withFullScreenPortal={false}
      withPortal={false}
    />
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Check-in Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T06:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="Check-out Time"
        onChange={[Function]}
        openTo="hours"
        value={"1970-01-01T02:00:00.000Z"}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <div
    className="date-time-picker"
  >
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="ETA"
        onChange={[Function]}
        openTo="hours"
        value={32400000}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
    <MuiPickersUtilsProvider
      utils={[Function]}
    >
      <PickerWithState
        ampm={true}
        invalidDateMessage="Invalid Time Format"
        label="ETD"
        onChange={[Function]}
        openTo="hours"
        value={0}
        views={
          Array [
            "hours",
            "minutes",
          ]
        }
      />
    </MuiPickersUtilsProvider>
  </div>
  <input
    className="text-input"
    onChange={[Function]}
    placeholder="location"
    type="text"
    value="Waikiki"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Add a note for your hotel (optional)"
    value=""
  />
  <button
    className="btn-push"
  >
    Save Hotel
  </button>
</form>
`;
